### PR TITLE
Crop to slide: support even cropping of non-picture objects #1179

### LINKED ISSUE #1179: Crop to slide: support even cropping of non-picture objects
I suppose we can convert the object to a picture first and then do the cropping?

### PR DESCRIPTION
**Pull request by jamos-tay**
Fixes #1179 

Cropping shapes now automatically casts them to picture.

Copy of #1186, rebased on alpha-release to clean branch history.

### CODE REVIEW

**Comment by damithc** [comment]
> It is possible for the cropped shape to remain as a shape if merge shapes is used.

IIRC, that api is not available in ppt 2010

**Comment by jamos-tay** [comment]
Ah, also I recall one of the requirements of CropToSlide was to set the picture's crop window rather than actually cropping the image. This allows the crop to be undone in future should the user want to.

**Comment by AngShiYa** [comment]
@damithc Yes, the api is not available but there is a workaround to execute merge shapes using the command bar. E.g. `this.ExecuteOfficeCommand("ShapesIntersect");`

@jamos-tay I see.. didn't know that requirement exist.

**Comment by damithc** [comment]
> Yes, the api is not available but there is a workaround to execute merge shapes using the command bar. E.g. this.ExecuteOfficeCommand("ShapesIntersect");

Nice. Are we using that already? I remember some features were done in a sub-optimal way simply because  of the 2010 API limitation. e.g. spotlight

**Comment by AngShiYa** [comment]
If I'm not wrong, we are not using merge shapes for any feature.

**Comment by jamos-tay** [comment]
Oh yeah, one thing to note is also that merge shapes uses the intersect of every selected shape, i.e. (A n B n C ...) so it wouldn't work for cases like this:

image

Since the intersection is an empty set.

**Comment by damithc** [comment]
> one thing to note is also that merge shapes uses the intersect of every selected shape, i.e. (A n B n C ...) so it wouldn't work for cases like this

On the UI, there are different options:

image

**Comment by jamos-tay** [comment]
> On the UI, there are different options:

I'm not sure how any of those help...

My point is that merge shapes wouldn't work for Crop to Slide when the user is cropping multiple shapes, for example a case like this:

image

We can't just create a shape of the slide and use Merge Shape because it uses the intersect of every single selected shape (A n B n C n ... n Slide), when actually what we want is the intersect of each shape with the slide ((A n Slide) u (B n Slide) u (C n Slide) u ...). We can get around this by creating a slide shape for each selected shape, but that seems rather inefficient.

File: PowerPointLabs/PowerPointLabs/CropLab/CropToSlide.cs [hunk 1]
```diff
@@ -81,5 +81,12 @@ private static RectangleF GetCropArea(PowerPoint.Shape shape, float slideWidth,
 
             return new RectangleF(cropLeft, cropTop, cropWidth, cropHeight);
         }
+
+        private static bool IsShape(PowerPoint.Shape shape)
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/CropLab/CropToSlide.cs]: Can move this to a util class?

File: PowerPointLabs/PowerPointLabs/CropLab/CropLabErrorHandler.cs [hunk 2]
```diff
@@ -55,6 +57,7 @@ public void ProcessErrorCode(int errorCode, string featureName, string validSele
                     }
                     ShowErrorMessage(errorCode, featureName, validSelectionMinCount.ToString(), validSelectionType);
                     break;
+                case ErrorCodeSelectionMustBeShapeOrPicture:
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/CropLab/CropLabErrorHandler.cs]: This switch case seems to be duplicated at line 60 and 112

Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/CropLab/CropLabErrorHandler.cs]: This switch case seems to be duplicated at line 60 and 112

File: PowerPointLabs/PowerPointLabs/CropLab/CropToSlide.cs (hunk 1 above)
Comment (jamos-tay) [inline PowerPointLabs/PowerPointLabs/CropLab/CropToSlide.cs]: Sure, I moved it to graphics

**Comment by damithc** [comment]
> We can't just create a shape of the slide and use Merge Shape because it uses the intersect of every single selected shape (A n B n C n ... n Slide), when actually what we want is the intersect of each shape with the slide ((A n Slide) u (B n Slide) u (C n Slide) u ...). We can get around this by creating a slide shape for each selected shape, but that seems rather inefficient.

Yes, no one option can give you the result you want. How about these 2 steps? 1. union all shapes 2. intersect with a shape that matches the slide

**Pull request merged**